File: calculate-csp-hashes.js
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// Function to calculate SHA-256 hash
function calculateHash(content) {
  return crypto.createHash('sha256').update(content, 'utf8').digest('base64');
}

// Function to extract and hash inline scripts and styles from HTML
function processHTMLFile(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  const hashes = { scripts: [], styles: [] };
  
  // Extract inline scripts
  const scriptMatches = content.matchAll(/<script[^>]*>([\s\S]*?)<\/script>/gi);
  for (const match of scriptMatches) {
    const scriptContent = match[1].trim();
    if (scriptContent && !match[0].includes('src=')) {
      const hash = calculateHash(scriptContent);
      hashes.scripts.push(`'sha256-${hash}'`);
      console.log(`Script hash for ${filePath}: sha256-${hash}`);
    }
  }
  
  // Extract inline styles
  const styleMatches = content.matchAll(/<style[^>]*>([\s\S]*?)<\/style>/gi);
  for (const match of styleMatches) {
    const styleContent = match[1].trim();
    if (styleContent) {
      const hash = calculateHash(styleContent);
      hashes.styles.push(`'sha256-${hash}'`);
      console.log(`Style hash for ${filePath}: sha256-${hash}`);
    }
  }
  
  return hashes;
}

// Find all HTML files
function findHTMLFiles(dir) {
  const files = [];
  const items = fs.readdirSync(dir, { withFileTypes: true });
  
  for (const item of items) {
    const fullPath = path.join(dir, item.name);
    if (item.isDirectory() && !item.name.startsWith('.') && item.name !== 'node_modules') {
      files.push(...findHTMLFiles(fullPath));
    } else if (item.isFile() && item.name.endsWith('.html')) {
      files.push(fullPath);
    }
  }
  
  return files;
}

// Main execution
const htmlFiles = findHTMLFiles('.');
const allScriptHashes = new Set();
const allStyleHashes = new Set();

console.log('Calculating CSP hashes for inline scripts and styles...\n');

for (const file of htmlFiles) {
  const hashes = processHTMLFile(file);
  hashes.scripts.forEach(h => allScriptHashes.add(h));
  hashes.styles.forEach(h => allStyleHashes.add(h));
}

console.log('\n=== CSP Configuration ===');
console.log('\nscript-src hashes:');
console.log(Array.from(allScriptHashes).join(' '));
console.log('\nstyle-src hashes:');
console.log(Array.from(allStyleHashes).join(' '));

File: calculate-csp-hashes.py
#!/usr/bin/env python3
"""
Calculate CSP hashes for inline scripts and styles in HTML files.
This script extracts all inline <script> and <style> tags and calculates SHA-256 hashes.
"""
import os
import re
import hashlib
import base64
from pathlib import Path

def calculate_hash(content):
    """Calculate SHA-256 hash and return base64 encoded string."""
    sha256 = hashlib.sha256(content.encode('utf-8')).digest()
    return base64.b64encode(sha256).decode('utf-8')

def process_html_file(file_path):
    """Extract and hash inline scripts and styles from an HTML file."""
    hashes = {'scripts': [], 'styles': []}
    
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception as e:
        print(f"Error reading {file_path}: {e}")
        return hashes
    
    # Extract inline scripts (not external scripts with src=)
    script_pattern = r'<script[^>]*>([\s\S]*?)</script>'
    for match in re.finditer(script_pattern, content, re.IGNORECASE):
        script_tag = match.group(0)
        script_content = match.group(1).strip()
        
        # Skip external scripts (those with src= attribute)
        if script_content and 'src=' not in script_tag:
            hash_value = calculate_hash(script_content)
            hash_string = f"'sha256-{hash_value}'"
            if hash_string not in hashes['scripts']:
                hashes['scripts'].append(hash_string)
                print(f"Script hash for {file_path}: sha256-{hash_value}")
    
    # Extract inline styles
    style_pattern = r'<style[^>]*>([\s\S]*?)</style>'
    for match in re.finditer(style_pattern, content, re.IGNORECASE):
        style_content = match.group(1).strip()
        if style_content:
            hash_value = calculate_hash(style_content)
            hash_string = f"'sha256-{hash_value}'"
            if hash_string not in hashes['styles']:
                hashes['styles'].append(hash_string)
                print(f"Style hash for {file_path}: sha256-{hash_value}")
    
    return hashes

def find_html_files(directory):
    """Recursively find all HTML files in a directory."""
    html_files = []
    for root, dirs, files in os.walk(directory):
        # Skip hidden directories and node_modules
        dirs[:] = [d for d in dirs if not d.startswith('.') and d != 'node_modules']
        
        for file in files:
            if file.endswith('.html'):
                html_files.append(os.path.join(root, file))
    
    return html_files

def main():
    """Main execution."""
    # Start from current directory
    base_dir = os.path.dirname(os.path.abspath(__file__)) or '.'
    
    print('Calculating CSP hashes for inline scripts and styles...\n')
    
    html_files = find_html_files(base_dir)
    all_script_hashes = set()
    all_style_hashes = set()
    
    for file_path in html_files:
        hashes = process_html_file(file_path)
        all_script_hashes.update(hashes['scripts'])
        all_style_hashes.update(hashes['styles'])
    
    print('\n=== CSP Configuration ===')
    print('\nscript-src hashes:')
    print(' '.join(sorted(all_script_hashes)))
    print('\nstyle-src hashes:')
    print(' '.join(sorted(all_style_hashes)))
    print()

if __name__ == '__main__':
    main()

